core/src/libraries/Oracle.sol

```diff
@@ -13,32 +15,74 @@ library Oracle {
+            //
+            //                                        i                   i-w
+            // 2T * (meanTick0To2W - meanTickWTo2W) = ∑ tick_n * dt_n  - 2∑ tick_n * dt_n
+            //                                        n=i-2w              n=i-2w
+            //
+            //                                        i                   i-w
+            //                                      = ∑ tick_n * dt_n  -  ∑ tick_n * dt_n
+            //                                        n=i-w               n=i-2w
+            //
+            // Thus far all values have been "true". We now assume that some manipulated value `manip_n` is added to each `tick_n`
+            //
+            //                                        i                               i-w
+            //                                      = ∑ (tick_n + manip_n) * dt_n  -  ∑ (tick_n + manip_n) * dt_n
+            //                                        n=i-w                           n=i-2w
+            //
+            //                                        i                   i-w                 i                    i-w
+            //                                      = ∑ tick_n * dt_n  -  ∑ tick_n * dt_n  +  ∑ manip_n * dt_n  -  ∑ manip_n * dt_n
+            //                                        n=i-w               n=i-2w              n=i-w                n=i-2w
+            //
+            //        meanTick0To2W - meanTickWTo2W = (meanTick0ToW_true - meanTickWTo2W_true) / 2  +  (sumManip0ToW - sumManipWTo2W) / (2T)
+            //
+            // For short time periods and reasonable market conditions, (meanTick0ToW_true - meanTickWTo2W_true) ≈ 0
+            //
+            //                                      ≈ (sumManip0ToW - sumManipWTo2W) / (2T)
+            //
+            // The TWAP we care about (see a few lines down) is measured over the interval [-w, 0). The result we've
+            // just derived contains `sumManip0ToW`, which is the sum of all manipulation in that same interval. As
+            // such, we use it as a metric for detecting manipulation. NOTE: If an attacker manipulates things to
+            // the same extent in the prior interval [-2w, -w), the metric will be 0. To guard against this, we must
+            // to watch the metric over the entire window. Even though it may be 0 *now*, it will have risen past a
+            // threshold at *some point* in the past `UNISWAP_AVG_WINDOW` seconds.
+            metric = uint56(SoladyMath.dist(meanTick0To2W, meanTickWTo2W));
 
-            sqrtMeanPriceX96 = TickMath.getSqrtRatioAtTick(arithmeticMeanTick);
-            secondsPerLiquidityX128 = secondsPerLiquidityCumulativeX128s[1] - secondsPerLiquidityCumulativeX128s[0];
+            // Compute arithmetic mean tick over `UNISWAP_AVG_WINDOW`, always rounding down to -inf
+            int256 delta = tickCumulatives[2] - tickCumulatives[1];
+            int256 meanTick0ToW = delta / int32(UNISWAP_AVG_WINDOW);
```

Comment:
$TWAP_{measured, -30mins \rightarrow now}$

This is the one we actually use as a price. Everything else is to determine whether this is manipulated or not.